### ISSUE TITLE
fermipy 1.0.0 GTAnalysis gta.setup() error

### ISSUE THREAD

**Issue by mealworm** [fermiPy/fermipy#369]
This is a copy of issue #365, but it applies to version 1.0.0 as well (which is supposed to work with fermitools 2.0.0). Basically there is still a deletion from a dictionary in use, which is an error in python3. The dictionary must be copied first, or some other strategy used.

from fermipy.gtanalysis import GTAnalysis
gta = GTAnalysis('config.yaml',logging={'verbosity': 3})
matplotlib.interactive(True)

WARNING: AstropyDeprecationWarning: astropy.extern.six will be removed in 4.0, use the six module directly if it is still needed [astropy.extern.six]
/srv/conda/envs/notebook/lib/python3.7/site-packages/fermipy/config.py:213: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  config_dict = yaml.load(open(config))
2020-10-12 08:12:52 INFO    GTAnalysis.__init__(): 
--------------------------------------------------------------------------------
fermipy version v1.0.0 
ScienceTools version 2.0.0
/srv/conda/envs/notebook/lib/python3.7/site-packages/fermipy/model_utils.py:59: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  get_function_spec.fndict = yaml.load(open(modelfile))
{'Prefactor': 0, 'Index1': 1, 'Scale': 2, 'Cutoff': 3, 'Index2': 4}
{'Prefactor': 0, 'Index1': 1, 'Scale': 2, 'Cutoff': 3, 'Index2': 4}
gta.setup()
2020-10-12 08:13:19 INFO    GTAnalysis.setup(): Running setup.
2020-10-12 08:13:19 INFO    GTBinnedAnalysis.setup(): Running setup for component 00
2020-10-12 08:13:19 INFO    GTBinnedAnalysis.run_gtapp(): Running gtselect.
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-1e27b2b06e27> in <module>
----> 1 gta.setup()

/srv/conda/envs/notebook/lib/python3.7/site-packages/fermipy/gtanalysis.py in setup(self, init_sources, overwrite, **kwargs)
   1080         # Run setup for each component
   1081         for i, c in enumerate(self.components):
-> 1082             c.setup(overwrite=overwrite)
   1083 
   1084         # Create likelihood

/srv/conda/envs/notebook/lib/python3.7/site-packages/fermipy/gtanalysis.py in setup(self, overwrite, **kwargs)
   5110         # Run data selection
   5111         if not use_external_srcmap:
-> 5112             self._select_data(overwrite=overwrite, **kwargs)
   5113 
   5114         # Create LT Cube

/srv/conda/envs/notebook/lib/python3.7/site-packages/fermipy/gtanalysis.py in _select_data(self, overwrite, **kwargs)
   5192 
   5193         run_gtapp('gtselect', self.logger, kw_gtselect,
-> 5194                   loglevel=loglevel)
   5195         if self.config['selection']['roicut'] == 'yes' or \
   5196                 self.config['selection']['filter'] is not None:

/srv/conda/envs/notebook/lib/python3.7/site-packages/fermipy/gtanalysis.py in run_gtapp(appname, logger, kw, **kwargs)
    199     logger.log(loglevel, 'Running %s.', appname)
    200     t0 = time.time()
--> 201     filter_dict(kw, None)
    202     kw = utils.unicode_to_str(kw)
    203     gtapp = GtApp.GtApp(str(appname))

/srv/conda/envs/notebook/lib/python3.7/site-packages/fermipy/gtanalysis.py in filter_dict(d, val)
    225 
    226 def filter_dict(d, val):
--> 227     for k, v in d.items():
    228         if v == val:
    229             del d[k]

RuntimeError: dictionary changed size during iteration

**Comment by cameronrulten** [comment]
I just tested this too and also get this setup error.

**Comment by Josef-prog-prog** [comment]
I encountered the same Runtime Error using the recent upgrade with python 3.7

**Comment by henrikef** [comment]
I believe `for k, v in list(d.items()):` should work with python3.

**Comment by mealworm** [comment]
> I believe `for k, v in list(d.items()):` should work with python3.

That part works, but not when there is 
del d[k]
in the loop -- that's where the error comes.

**Comment by henrikef** [comment]
Let me try again.

If you replace the loop

```
for k, v in d.items():
``` 
with 

```
for k, v in list(d.items()):
```
 
, then the `del d[k]` command should work without crashing in python3.

**Comment by Josef-prog-prog** [comment]
> Let me try again.
> 
> If you replace the loop
> 
> ```
> for k, v in d.items():
> ```
> 
> with
> 
> ```
> for k, v in list(d.items()):
> ```
> 
> , then the `del d[k]` command should work without crashing in python3.

it works, thank you. But, there is a key error while running the light curve with the message pasted below;
  --------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-22-f7ba32602b70> in <module>()
----> 1 lc = gta.lightcurve('4FGL J1800.6+7828', free_params=['Prefactor'], binsz=43200, make_plots=True)   #86400.*15.0days bins    #time_bins=binsLC)

/home/josef/miniconda3/envs/fermi/lib/python3.7/site-packages/fermipy/lightcurve.py in lightcurve(self, name, **kwargs)
    287         self.logger.info('Computing Lightcurve for %s' % name)
    288 
--> 289         o = self._make_lc(name, **config)
    290         filename = utils.format_filename(self.workdir, 'lightcurve',
    291                                          prefix=[config['prefix'],

/home/josef/miniconda3/envs/fermi/lib/python3.7/site-packages/fermipy/lightcurve.py in _make_lc(self, name, **kwargs)
    450             next_fit = next(mapo)
    451 
--> 452             if not next_fit['fit_success']:
    453                 self.logger.error(
    454                     'Fit failed in bin %d in range %i %i.' % (i, time[0], time[1]))

KeyError: 'fit_success'

**Comment by dieterhorns** [comment]
Following up on this - I have had the same problem, solved it with 
```python
dcopy={**d}
for k,v in dcopy.items():
    if v == val:
       del d[k]
```
following the recommendation not to iterate over a dictionary and trying to change it.

**Comment by PhysicistSpock** [comment]
> Following up on this - I have had the same problem, solved it with
> 
> ```python
> dcopy={**d}
> for k,v in dcopy.items():
>     if v == val:
>        del d[k]
> ```
> 
> following the recommendation not to iterate over a dictionary and trying to change it.

Unfortunately tryinf this and previous comments did not wokr for me. I still get a "RuntimeError: dictionary changed size during iteration". 
Is there any update on this?

**Comment by grburgess** [comment]
Is there any update on this? Currently it is impossible to run an analysis.

**Comment by eacharles** [comment]
Fixed with #380.

**Comment by sstraal** [comment]
I was also running into this issue on MacOS, using the FermiBottle (docker) and the solutions of @dieterhorns and @henrikef did not work for me. And hence the solution fixed in #380 also did not work for me (as it only incorporates one of the solutions).

However, if you combine both solutions, it does work. 
See the solution found by @jeagle94 
https://github.com/fermi-lat/FermiBottle/issues/10#issuecomment-797065594
